tidb_binlog/pump_client/client.go

```diff
@@ -0,0 +1,393 @@
+		selector = NewHashSelector()
+	case Score:
+		selector = NewScoreSelector()
+	default:
+		log.Warnf("unknow algorithm %s, use hash as default", algorithm)
+		selector = NewHashSelector()
+	}
+
+	ectdEndpoints, err := utils.ParseHostPortAddr(etcdURLs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cli, err := etcd.NewClientFromCfg(ectdEndpoints, DefaultEtcdTimeout, RootPath, security)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	newPumpsClient := &PumpsClient{
+		ctx:                ctx,
+		cancel:             cancel,
+		EtcdCli:            cli,
+		Selector:           selector,
+		RetryTime:          DefaultRetryTime,
+		BinlogWriteTimeout: DefaultBinlogWriteTimeout,
+	}
+
+	err = newPumpsClient.getPumpStatus(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	newPumpsClient.Selector.SetPumps(newPumpsClient.AvaliablePumps)
+
+	newPumpsClient.wg.Add(2)
+	go newPumpsClient.watchStatus()
+	go newPumpsClient.heartbeat()
+
+	return newPumpsClient, nil
+}
+
+// getPumpStatus retruns all the pumps status in the etcd.
+func (c *PumpsClient) getPumpStatus(pctx context.Context) error {
+	c.Lock()
+	defer c.Unlock()
+
+	ctx, cancel := context.WithTimeout(pctx, DefaultEtcdTimeout)
+	defer cancel()
+
+	resp, err := c.EtcdCli.List(ctx, path.Join(RootPath))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	nodesStatus, err := node.NodesStatusFromEtcdNode(resp)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, status := range nodesStatus {
+		log.Infof("get pump %v from etcd", status)
+		c.addPump(status)
+	}
+
+	return nil
+}
+
+// WriteBinlog writes binlog to a situable pump.
+func (c *PumpsClient) WriteBinlog(clusterID uint64, binlog *pb.Binlog) error {
+	c.RLock()
+	pump := c.Selector.Select(binlog)
+	c.RUnlock()
+	log.Infof("write binlog choose pump %v", pump)
+
+	commitData, err := binlog.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	req := &pb.WriteBinlogReq{ClusterID: clusterID, Payload: commitData}
+
+	// Retry many times because we may raise CRITICAL error here.
+	for i := 0; i < c.RetryTime; i++ {
+		if pump == nil {
+			return errors.New("no pump can use")
+		}
+
+		resp, err := c.writeBinlog(req, pump)
+		if err == nil && resp.Errmsg != "" {
+			err = errors.New(resp.Errmsg)
+		}
+		if err == nil {
+			return nil
+		}
+		if strings.Contains(err.Error(), "received message larger than max") {
+			// This kind of error is critical and not retryable, return directly.
+			return errors.Errorf("binlog data is too large (%s)", err.Error())
+		}
+
+		log.Errorf("write binlog error %v", err)
+
+		// every pump can retry 5 times, if retry 5 times and still failed, set this pump unavaliable, and choose a new pump.
+		if (i+1)%5 == 0 {
+			c.Lock()
+			c.setPumpAvaliable(pump, false)
+			c.Unlock()
+			pump = c.Selector.Next(pump, binlog, i/5+1)
+			log.Infof("avaliable pumps: %v, write binlog choose pump %v", c.AvaliablePumps, pump)
```

Comment:
when i = c.RetryTime, you choose one pump to write, and log say you are going to write this pump, but it will quit the for loop right away, this looks confusing